lib/cram.ml

```diff
@@ -67,38 +52,40 @@ let pp ?pad ppf (t : t) =
   pp_lines (Output.pp ?pad) ppf t.output;
   pp_exit_code ?pad ppf t.exit_code
 
-let hpad_of_lines = function
+let rec hpad_of_lines = function
   | [] -> 0
-  | h :: _ ->
-      let i = ref 0 in
-      while !i < String.length h && h.[!i] = ' ' do
-        incr i
-      done;
-      !i
+  | h :: hs -> (
+      match String.equal String.empty h with
+      | true -> hpad_of_lines hs
+      | false ->
+          let i = ref 0 in
+          while !i < String.length h && h.[!i] = ' ' do
+            incr i
+          done;
+          !i)
 
-let of_lines ~syntax ~(loc : Location.t) t =
-  let pos = loc.loc_start in
-  let hpad =
-    match syntax with Syntax.Mli -> pos.pos_cnum + 2 | _ -> hpad_of_lines t
-  in
+let is_whitespace s = s |> String.trim |> String.is_empty
+
+let of_lines t =
+  let hpad = hpad_of_lines t in
   let unpad line =
-    match syntax with
-    | Syntax.Mli -> String.trim line
-    | _ ->
-        if String.is_empty line then line
-        else if String.length line < hpad then
-          Fmt.failwith "invalid padding: %S" line
-        else String.with_index_range line ~first:hpad
+    match is_whitespace line with
+    | true -> String.with_index_range line ~first:hpad
+    | false -> (
+        match String.length line < hpad with
+        | true -> Fmt.failwith "invalid padding: %S" line
+        | false -> String.with_index_range line ~first:hpad)
   in
   let lines = List.map unpad t in
+  (* remove leading empty line, if any which is true for multipline blocks *)
+  let lines = match lines with "" :: lines -> lines | lines -> lines in
```

Comment:
This code is only used to remove the initial empty line that is created in multiline blocks (since the first line starts right at the beginning of a block and usually is just followed by a line break before the "proper" indented block starts (which is also why `hpad_of_lines` recurses, because the first line might be an outlier and not have any `hpad` at all)).

Now that you point it out, it is somewhat of an ugly hack because the cram parser rejects it otherwise but I feel like this might be the wrong approach. It seems to lead to the issue https://github.com/realworldocaml/mdx/pull/397/files#diff-406c39c844335a646d3100b6fba8ae3b2add673571869dd2a7315ba00b3e03f9R100 where a `\n` needs to be conditionally added because we removed it here and lost the information whether it was there or not.

Maybe we should just not mess with the empty lines and preserve them? What do you think?

Reply:
I would say that the least we have to have special cases with blank lines, the best.
So, I think it is better to keep the blank lines, and modify `lexer_cram.mll` to ignore leading whitespace. Not sure though!

```suggestion
```

Reply:
It's a good point. Upon investigation the tokenizer just parses those as `` `Output ""`` so I wrote some code to remove them and convert them to `vpad`. Then I also removed some more special casing where depending on whether it is in a cram or `.mli` file it would output empty outputs in a padded or non-padded way. It's somewhat of an edge-case but I think being consistent and predictable trumps avoiding trailing spaces.